### PR TITLE
txscript: Add null data script creator

### PR DESCRIPTION
**Pull request by DanielKrawisz**
NullDataScript takes a byte array and constructs a null data output script. This is used for constructing transactions that contain extra data in provably prunable outputs. 

There is also a bug fix in isNullData.

### CODE REVIEW

File: txscript/standard.go [hunk 1]
```diff
@@ -350,6 +350,21 @@ func PayToAddrScript(addr btcutil.Address) ([]byte, error) {
...
 }
 
+// NullDataScript creates a provably prunable script
+// containing OP_RETURN followed by the passed data.
+func NullDataScript(data []byte) ([]byte, error) {
+	if len(data) > MaxDataCarrierSize {
+		return nil, ErrStackLongScript
+	}
+
+	script := make([]byte, 3+len(data))
```
Comment (davecgh) [inline txscript/standard.go]: This will produce OP_RETURN scripts with non-canonical data pushes that would end up being rejected by the majority of the network.

Please use this instead:

````Go
	builder := txscript.NewScriptBuilder()
	return builder.AddOp(txscript.OP_RETURN).AddData(data).Script()
```

File: txscript/standard_test.go [hunk 2]
```diff
@@ -1013,3 +1013,28 @@ func TestStringifyClass(t *testing.T) {
 		}
 	}
 }
+
+// TestNullDataScript tests whether NullDataScript returns a valid script.
+func TestNullDataScript(t *testing.T) {
+	data := []byte{0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,
```
Comment (davecgh) [inline txscript/standard_test.go]: Please make this table driven so different data can be tested.

Something like the following

```Go
tests := []struct{
	name string
	data []byte
	valid bool
	expected []byte
}{
	{
		name: "more than max allowed",
		data: hexToBytes("000102....80"),
		valid: false,
		expected: nil,
	},
	{
		name: "small data",
		data: hexToBytes("01"),
		valid: false,
		expected: []byte{OP_RETURN, 0x51},
	},
	...
}

for i, test := range tests {
	... // Call func, check for err against valid, bytes.Compare(result, test.expected), etc
}
````

File: txscript/standard.go (hunk 1 above)
Comment (DanielKrawisz) [inline txscript/standard.go]: Sorry, but what's noncononical about what I did?

Comment (davecgh) [inline txscript/standard.go]: Another case would be:

`txscript.NullDataScript(bytes.Repeat([]byte{0x01, 24))`

Your code would produce `OP_RETURN OP_PUSHDATA1 0x18 0x01010101...` instead of `OP_RETURN OP_DATA_24 0x01010101...`.

File: txscript/standard_test.go (hunk 2 above)
Comment (davecgh) [inline txscript/standard_test.go]: Another test:

    {
        name: "small data",
        data: hexToBytes("0102030405060708090a0b0c0d0e0f101112131415161718"),
        valid: false,
        expected: hexToBytes("6a180102030405060708090a0b0c0d0e0f101112131415161718"),
    },

File: txscript/standard.go (hunk 1 above)
Comment (DanielKrawisz) [inline txscript/standard.go]: I guess I just don't understand bitcoin scripts well enough.

Comment (davecgh) [inline txscript/standard.go]: In case he above wasn't clear, here is some code using both that should illustrate better:

```Go
func main() {
	data := hexToBytes("0102030405060708091a1b1c1d1e1f101112")
	script, _ := txscript.NewScriptBuilder().AddOp(txscript.OP_RETURN).AddData(data).Script()
	disasm, _ := txscript.DisasmString(script)
	fmt.Printf("%x\n", script)
	fmt.Println(disasm)

	script, _= NullDataScript(data)
	disasm, _ = txscript.DisasmString(script)
	fmt.Printf("%x\n", script)
	fmt.Println(disasm)}
```

Output:
```
6a120102030405060708091a1b1c1d1e1f101112
OP_RETURN 0102030405060708091a1b1c1d1e1f101112
6a4c120102030405060708091a1b1c1d1e1f101112
OP_RETURN 0102030405060708091a1b1c1d1e1f101112
```

Notice how they both disassemble to the same thing, but the script builder version is the shorter, canonical form.

File: txscript/standard_test.go (hunk 2 above)
Comment (DanielKrawisz) [inline txscript/standard_test.go]: Thanks for all your advice but I'm still pretty confused. Your example "01" => "6a51" is what the code you suggested produces, but the result is not recognized as a NullDataTy when I use GetScriptClass. It returns NonStandardTy. Is that correct? If so, then NullDataScript should probably return an error rather than 6a51.

Comment (davecgh) [inline txscript/standard_test.go]: Nice.  We just found a bug in the `GetScriptClass` function.  I confirmed that 6a51 is indeed showing up as `nonstandard`, but it most certainly is a `nulldata` script.

```
{
  "asm": "OP_RETURN 1",
  "type": "nulldata",
  "p2sh": "35zfU45S1vbHDhfY3ibN8XgCafBUzTvT3a"
}
```

Comment (DanielKrawisz) [inline txscript/standard_test.go]: Great, well that would explain why my original function was wrong because I just followed the existing isNullData function to make something that would fit it. I have updated my pr, and one of the tests doesn't pass yet because of this bug. It will pass once isNullData is fixed though.

**Comment by davecgh** [comment]
I pushed #790 to fix the `GetScriptClass` nulldata with small integer detection issue discussed in the review.

**Comment by DanielKrawisz** [comment]
I thiiiink I figured out how to correct isNullData and I added a new test case to scriptClassTests. 

As for whether this pr is necessary, all I wanted to do was add an OP_RETURN output to a tx and that was a pretty big job for someone who is just learning the code base! It sure would have been convenient for me to have had this func available from the start.

File: txscript/standard_test.go [hunk 3]
```diff
@@ -1013,3 +1019,89 @@ func TestStringifyClass(t *testing.T) {
...
+
+	tests := []struct {
+		name     string
+		data     []byte
+		valid    bool
+		expected []byte
+	}{
+		{
+			name:     "small data",
+			data:     tiny,
```
Comment (davecgh) [inline txscript/standard_test.go]: You can use `data: hexToBytes("01"),` to do the scripts inline.

File: txscript/standard_test.go [hunk 4]
```diff
@@ -1013,3 +1019,89 @@ func TestStringifyClass(t *testing.T) {
...
+		if err != nil {
+			t.Errorf("NullDataScript test %v: Unexpected error returned %v ",
+				test.name, err)
+
+			continue
+		}
+
+		// Check that the expected result was returned.
+		if !bytes.Equal(script, test.expected) {
+			t.Errorf("NullDataScript test %v: Expected %v, got %v",
```
Comment (davecgh) [inline txscript/standard_test.go]: You can use %x in the `Errorf` format string to get hex.

```Go
t.Errorf("%s: expected %x, got %x`, test.name, test.expected, script)
```

Also, as above, please include a `continue` after the `t.Errorf` so it continues to the next test.

File: txscript/standard_test.go [hunk 5]
```diff
@@ -1013,3 +1019,89 @@ func TestStringifyClass(t *testing.T) {
...
+			name:     "too big",
+			data:     tooLong,
+			valid:    false,
+			expected: nil,
+		},
+	}
+
+	for _, test := range tests {
+		script, err := txscript.NullDataScript(test.data)
+
```
Comment (davecgh) [inline txscript/standard_test.go]: Please combine the checks here and add a continue after the error so it skips to the next test:

```Go
if test.valid && err != nil {
	t.Errorf("....")
	continue
} else if !test.valid && err == nil {
	t.Errorf("....")
	continue
}
```

File: txscript/standard.go [hunk 6]
```diff
@@ -132,14 +132,23 @@ func isNullData(pops []parsedOpcode) bool {
...
 	}
 
+	// Otherwise, OP_RETURN followed by some data is allowed.
+	// This can either be in the form of an OP_PUSHDATA, an
+	// OP_DATA, or an OP_0 through OP_16.
 	return l == 2 &&
 		pops[0].opcode.value == OP_RETURN &&
-		pops[1].opcode.value <= OP_PUSHDATA4 &&
-		len(pops[1].data) <= MaxDataCarrierSize
+		((pops[1].opcode.value <= OP_PUSHDATA4 &&
```
Comment (davecgh) [inline txscript/standard.go]: I pushed PR #709 which does this via `isSmallInt` which is designed for this purpose and adds a bunch of tests for the various edge conditions.  Please remove this and rebase the PR on top of 709.

**Comment by davecgh** [comment]
Yes, I agree it's more convenient to do:

```Go
script, err := txscript.NullDataScript(data)
```
versus
```Go
script, err := txscript.NewScriptBuilder().AddOp(txscript.OP_RETURN).AddData(data).Script()
```

I appreciate the PR.

File: txscript/standard.go [hunk 7]
```diff
@@ -350,6 +359,17 @@ func PayToAddrScript(addr btcutil.Address) ([]byte, error) {
...
 
+// NullDataScript creates a provably prunable script
+// containing OP_RETURN followed by the passed data.
+func NullDataScript(data []byte) ([]byte, error) {
+	if len(data) > MaxDataCarrierSize {
+		return nil, ErrStackLongScript
+	}
+
+	builder := NewScriptBuilder()
+	return builder.AddOp(OP_RETURN).AddData(data).Script()
```
Comment (davecgh) [inline txscript/standard.go]: I know I originally showed this on two lines, but I forgot it was in the `txscript` package itself, so it'll fit on a single line.

Mind making it the following single line?

```Go
return NewScriptBuilder().AddOp(OP_RETURN).AddData(data).Script()
```

File: txscript/standard_test.go (hunk 3 above)
Comment (DanielKrawisz) [inline txscript/standard_test.go]: standard_test.go is part of package txscript_test, so hexToBytes isn't available, as it is in package txscript.

Comment (davecgh) [inline txscript/standard_test.go]: Ok, I see it's `decodeHex` in the test package.  I'll need to combine those when doing aforementioned PR.

File: txscript/standard_test.go (hunk 5 above)
Comment (DanielKrawisz) [inline txscript/standard_test.go]: Ok, I did this but I think it makes the rest of the test look a little weird. I updated the pr so you can see for yourself.

**Comment by davecgh** [comment]
Hmm, not sure how you rebased it there, but it somehow ended up with an extra `btcjson` commit.

There should only end up being two commits:
`* txscript: Correct nulldata standardness check.`  (the one from PR #790)
`* txscript: Add NullDataScript function.`  (your commit)

The git command should look something like:

```bash
$ git rebase -i --onto 4d4612 f78f2fb^
```

Where 4d4612 is the `txscript: Correct nulldata standardness check` commit and f78f2fb is your first commit that you want to squash into.

**Comment by davecgh** [comment]
Both #790 and #791 have been merged to master now, so you should just be able to do a normal rebase to master now.

**Comment by DanielKrawisz** [comment]
Ok I fixed the rebasing... I don't know how that happened, sorry!
